### PR TITLE
KAFKA-5260: Producer should not send AbortTxn unless transaction has actually begun

### PR DESCRIPTION
**Pull request by dguy**
Keep track of when a transaction has begun by setting a flag, `transactionStarted` when a successfull `AddPartitionsToTxnResponse` or `AddOffsetsToTxnResponse` had been received. If an `AbortTxnRequest` about to be sent and `transactionStarted` is false, don't send the request and transition the state to `READY`

### CODE REVIEW

**Comment by dguy** [comment]
@apurvam @hachikuji - This seems to work, though i'm not entirely sure it is correct!

File: clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java [hunk 1]
```diff
@@ -624,6 +624,68 @@ public void shouldNotAddPartitionsToTransactionWhenTopicAuthorizationFailed() th
...
+        transactionManager.maybeAddPartitionToTransaction(tp0);
+
+        TransactionalRequestResult abortResult = transactionManager.beginAbortingTransaction();
+
+        prepareAddPartitionsToTxnResponse(Errors.TOPIC_AUTHORIZATION_FAILED, tp0, epoch, pid);
+        sender.run(time.milliseconds());  // Send AddPartitionsRequest
+        assertFalse(abortResult.isCompleted());
+
+        sender.run(time.milliseconds());
+        assertTrue(abortResult.isCompleted());
```
Comment (apurvam) [inline clients/src/test/java/org/apache/kafka/clients/producer/internals/TransactionManagerTest.java:766]: shall we also assert that the result is successful?

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala [hunk 2]
```diff
@@ -885,8 +885,10 @@ class AuthorizerIntegrationTest extends BaseRequestTest {
   def shouldThrowTransactionalIdAuthorizationExceptionWhenNoTransactionAccessOnEndTransaction(): Unit = {
     addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Write)), producerTransactionalIdResource)
     addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Write)), Resource.ProducerIdResource)
+    addAndVerifyAcls(Set(new Acl(KafkaPrincipal.ANONYMOUS, Allow, Acl.WildCardHost, Write)), topicResource)
```
Comment (apurvam) [inline core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: Should we also add a test case where there is a real authorization failure and yet the transaction can be aborted?

**Review by apurvam — COMMENTED** [review]
The changes look correct. However, I am not a fan of the one off `transactionStarted` variable. Wouldn't it be nicer to split the `IN_TRANSACTION` state into `STARTING_TRANSACTION` and `STARTED_TRANSACTION`? You will only transition to the latter from the `STARTING_STATE` if you get a successful response from the broker. 

Then, if you get a commit/abort in `STARTING_STATE` you can handle it without an RPC. 

This works within the existing state model and will be easier to reason about going forward.

**Comment by dguy** [comment]
@apurvam i did start adding another state as you've suggested, though it seemed to be getting pretty messy due to having to check if the currentState is either `STARTED_TRANSACTION` or `IN_TRANSACTION` in multiple places. I'll have another look, but it wasn't really filling me with joy hence the boolean

**Comment by dguy** [comment]
@apurvam i don't think this works by adding another state. The state transitions are happening on the Producers thread, yet we don't know we have successfully completed a transactional request until we get an `AddPartitionsToTxnResponse` or an `AddOffsetsToPartitionResponse`, which is happening on the sender thread. In the meantime the current state may have transitioned to subsequent states _before_ we get these responses. So when it comes to the time when we need to decide if we should send the `EndTxnRequest` we don't have any idea if the transaction was previously in the `STARTING_TRANSACTION` state. I thought adding a `previousState` might help, but no. As there may have been multiple transitions

File: core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala (hunk 2 above)
Comment (dguy) [reply core/src/test/scala/integration/kafka/api/AuthorizerIntegrationTest.scala]: By authorization failure do you mean topic authorization?

**Comment by apurvam** [comment]
@dguy I started typing a really long response, but deleted it. I guess the key point is that if we never transitioned from `STARTING_TRANSACTION` to `IN_TRANSACTION`  at the time of commit or abort, it means that either there are no `send` or `sendOffset` calls attempted, or they have failed. In the former, the additional state will do what we want: if we are committing or aborting in this state, we know there is nothing to do. 

HOwever, if the `AddPartitions` or `AddOffsets` RPCs failed fatally, or have never been successfully acknowledged, we would --correctly-- be in an error state, and hence have no record of whether we ever got out of `STARTING_TRANSACTION` to `IN_TRANSACTION`. 

So either we hack the transitions to the error state, or just maintain the separate variable. I think the separate variable is preferable.

File: clients/src/main/java/org/apache/kafka/clients/NetworkClient.java [hunk 3]
```diff
@@ -750,6 +750,8 @@ public void handleCompletedMetadataResponse(RequestHeader requestHeader, long no
             Cluster cluster = response.cluster();
             // check if any topics metadata failed to get updated
             Map<String, Errors> errors = response.errors();
+            metadata.lastErrors(errors);
```
Comment (apurvam) [inline clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: Was this required to get your authorization integration test to work? If so, why?

Comment (dguy) [reply clients/src/main/java/org/apache/kafka/clients/NetworkClient.java]: Oops, that was me debugging. Will remove it

**Comment by guozhangwang** [comment]
Actually, it seems that currently `FATAL_ERROR` and `ABORTABLE_ERROR` states are not distinguished. I'm wondering if we can just use `FATAL_ERROR` state and in that state do not send abortTxn?

**Comment by hachikuji** [comment]
@guozhangwang The difference between `FATAL_ERROR` and `ABORTABLE_ERROR` is that the former state will throw an exception back at the user even if they try to call `abortTransaction()`. The only possible recovery is closing the producer (see related issue https://issues.apache.org/jira/browse/KAFKA-5342). In some cases, we may reach an abortable error before any partition has been successfully added to the transaction (this is unfortunately the consequence of not having a BeginTxn request). We want the user to call `abortTransaction()` in this case, but we want to avoid sending EndTxn since this would cause an invalid transaction state.

I've been thinking about the current patch for a while and finally it LGTM. The subtle thing is the fact that we'll keep retrying AddPartition failures as long as we get retriable errors. If we receive an abortable error, we'll stop retrying and the EndTxn should reach the head of the queue. Luckily we made the AddPartitions request have all or nothing semantics, so any failure would mean we haven't started the transaction (assuming there were no previous successes).

**Pull request closed without merging**